perf(rolldown): minor performance optimization

Brooooooklyn opened the pull request:
- pre-allocate memory
- use simdutf8 rather than String utf8 check
- avoid double utf8 checks in text_to_string_literal

Brooooooklyn commented:
* **#6306** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/6306?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#6305** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/6305?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#6272** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/6272?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#6271** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/6271?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/6271?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#6270** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/6270?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add the label <kbd>graphite: merge</kbd> to this PR to add it to the merge queue.

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/rolldown?ref=merge-queue-instructions-in-stack-comment&prId=6271).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Brooooooklyn commented on crates/rolldown_plugin_data_uri/src/lib.rs:
```diff
@@ -48,7 +48,8 @@ impl Plugin for DataUriPlugin {
         let data = base64_simd::STANDARD.decode_to_vec(parsed.data)?;
-        String::from_utf8_lossy(data.as_ref()).as_ref().into()
+        // SAFETY: `data` is decoded by base64-simd and is valid utf8
+        unsafe { String::from_utf8_unchecked(data) }.into()
```
You are right

Brooooooklyn commented:
cursor review

IWANABETHATGUY commented:
### Merge activity

* **Sep 18, 5:12 PM UTC**: The merge label 'graphite: merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=rolldown&repo=rolldown) once it meets the requirements.
* **Sep 18, 5:12 PM UTC**: `IWANABETHATGUY` added this pull request to the [Graphite merge queue](https://app.graphite.dev/merges?org=rolldown&repo=rolldown).
* **Sep 18, 5:33 PM UTC**: Merged by the [Graphite merge queue](https://app.graphite.dev/merges?org=rolldown&repo=rolldown).

Brooooooklyn committed:
perf(rolldown): minor performance optimization (#6271)

- pre-allocate memory
- use simdutf8 rather than String utf8 check
- avoid double utf8 checks in text_to_string_literal
